### PR TITLE
Add helper for system.SystemRule and flow.FlowRule in datasource

### PR DESCRIPTION
**Pull request by louyuting**
### Describe what this PR does / why we need it
Encapsulate converter(default json) and updater related logic for specific property.
properties:

1. []system.SystemRule
2. []flow.FlowRule

### Special notes for reviews

### CODE REVIEW

File: ext/datasource/plugin/flow_rule.go [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+	rules := make([]*flow.FlowRule, 0)
+	if data == nil {
+		rules = nil
+	} else {
+		val, ok := data.([]flow.FlowRule)
+		if !ok {
+			return errors.New(fmt.Sprintf("Fail to type assert data to []flow.FlowRule, in fact, data: %+v", data))
+		}
+		for _, v := range val {
+			rules = append(rules, &v)
```
Comment (wenxuwan) [inline ext/datasource/plugin/flow_rule.go]: why not use val directly ?

Comment (louyuting) [reply ext/datasource/plugin/flow_rule.go]: Avoid redundant assignment copies.

File: ext/datasource/plugin/flow_rule.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+	return ret, nil
+}
+
+// FlowRulesUpdater load the newest []flow.FlowRule to downstream flow component.
+func FlowRulesUpdater(data interface{}) error {
+	rules := make([]*flow.FlowRule, 0)
+	if data == nil {
+		rules = nil
+	} else {
+		val, ok := data.([]flow.FlowRule)
```
Comment (sczyh30) [inline ext/datasource/plugin/flow_rule.go]: Maybe here we could accept both `[]flow.FlowRule` and `[]*flow.FlowRule` type?

File: ext/datasource/plugin/flow_rule.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/alibaba/sentinel-golang/ext/datasource"
+	"github.com/pkg/errors"
+)
+
+// FlowRulesJsonConverter provide JSON  as the default serialization for flow.FlowRule
+func FlowRulesJsonConverter(src []byte) (interface{}, error) {
+	if len(src) == 0 {
+		return nil, nil
+	}
+	ret := make([]flow.FlowRule, 0)
```
Comment (sczyh30) [inline ext/datasource/plugin/flow_rule.go]: Maybe we could use `[]*flow.FlowRule` here?

File: ext/datasource/plugin/flow_rule.go [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+		val, ok := data.([]flow.FlowRule)
+		if !ok {
+			return errors.New(fmt.Sprintf("Fail to type assert data to []flow.FlowRule, in fact, data: %+v", data))
+		}
+		for _, v := range val {
+			rules = append(rules, &v)
+		}
+	}
+	succ, err := flow.LoadRules(rules)
+	if succ {
```
Comment (sczyh30) [inline ext/datasource/plugin/flow_rule.go]: The `bool` returned by `LoadRules()` does not indicate error, so maybe we could just return `err` here?

File: core/flow/rule_manager.go [hunk 5]
```diff
@@ -80,6 +80,13 @@ func GetRules() []*FlowRule {
 	return rulesFrom(tcMap)
 }
 
+func ResetTcMap() {
```
Comment (sczyh30) [inline core/flow/rule_manager.go]: Use `LoadRules(nil)` instead?

Comment (sczyh30) [reply core/flow/rule_manager.go]: We may not need such `reset` functions actually :)
If necessary, it should be `ClearRules()`.

Comment (louyuting) [reply core/flow/rule_manager.go]: ClearRules() seem make more sense.
What you see is what you get.

File: ext/datasource/plugin/flow_rule.go (hunk 4 above)
Comment (louyuting) [reply ext/datasource/plugin/flow_rule.go]: make sense

File: ext/datasource/plugin/flow_rule.go (hunk 2 above)
Comment (louyuting) [reply ext/datasource/plugin/flow_rule.go]: Make sense

File: core/flow/rule_manager.go [hunk 6]
```diff
@@ -80,6 +80,15 @@ func GetRules() []*FlowRule {
 	return rulesFrom(tcMap)
 }
 
+func ClearRules() error {
+	isUpdated, err := LoadRules(nil)
+	if isUpdated {
+		return err
+	} else {
+		return errors.Wrap(err, "Fail to clear rules.")
```
Comment (sczyh30) [inline core/flow/rule_manager.go]: This does not indicate failure, as the rules may have been cleared before. Maybe here we could just return `err`?

File: core/system/rule_manager.go [hunk 7]
```diff
@@ -40,6 +40,14 @@ func initRuleRecvTask() {
 		}
 	}, logger)
 }
+func ClearRules() error {
+	isUpdated, err := LoadRules(nil)
+	if isUpdated {
+		return err
+	} else {
+		return errors.Wrap(err, "Fail to clear rules.")
```
Comment (sczyh30) [inline core/system/rule_manager.go]: Ditto.

File: go.mod [hunk 8]
```diff
@@ -14,4 +14,5 @@ require (
 	golang.org/x/sys v0.0.0-20200107162124-548cf772de50 // indirect
 	gopkg.in/yaml.v2 v2.2.2
 	go.uber.org/multierr v1.5.0
+	github.com/fsnotify/fsnotify v1.4.7
```
Comment (sczyh30) [inline go.mod]: Is it necessary?

Comment (louyuting) [reply go.mod]: I will remove it.
Thanks

**Comment by sczyh30** [comment]
Thanks!

**Pull request merged**